From 375703df3e5fedc60899bfe9e651f2ad0a9b47ce Mon Sep 17 00:00:00 2001
From: Lindsey Gray <lindsey.gray@gmail.com>
Date: Mon, 30 Jun 2025 13:12:40 -0500
Subject: [PATCH] changes to enable a python-extension-only build of fastjet

---
 CMakeLists.txt             | 92 ++++++++++++++++++++++----------------
 pyinterface/CMakeLists.txt |  9 ++--
 2 files changed, 59 insertions(+), 42 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index ed194c68..b8058c1e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -188,6 +188,7 @@ option(FASTJET_HAVE_EXPLICIT_FOR_OPERATORS  "Compiler supports explicit operator
 
 # support for swig-based python bindings
 option(FASTJET_ENABLE_PYTHON "Enable building of swig-based fastjet python bindings" OFF)
+option(FASTJET_PYINTERFACE_ONLY "Only build the pyinterface model, relying on an external install of fastjet" OFF)
 
 #-------------------------------------------------------------------------
 # compile-time options
@@ -336,19 +337,29 @@ set(CONFIG_SUMMARY "FastJet configuration summary:
 
 #-------------------------------------------------------------------------
 # get into directories:
-add_subdirectory(include)  # for basic FJ headers
-add_subdirectory(src)      # for the main FJ library
-add_subdirectory(tools)    # fastjet toolkit
-add_subdirectory(plugins)  # fastjet clustering plugins
+if (NOT FASTJET_PYINTERFACE_ONLY)
+  add_subdirectory(include)  # for basic FJ headers
+  add_subdirectory(src)      # for the main FJ library
+  add_subdirectory(tools)    # fastjet toolkit
+  add_subdirectory(plugins)  # fastjet clustering plugins
+endif()
 message(STATUS "Installation prefix is ${CMAKE_INSTALL_PREFIX}")
 if (FASTJET_ENABLE_PYTHON)
+  if (FASTJET_PYINTERFACE_ONLY)
+    find_package(fastjet REQUIRED)
+    if(fastjet_FOUND)
+      message(STATUS "Found fastjet for pyinterface-only build: ${fastjet_DIR}")    
+    endif()
+  endif()
   add_subdirectory(pyinterface) # for python interface
   if (NOT IS_ABSOLUTE "${FASTJET_PYTHON_INSTALL_DIR}")
     set(FASTJET_PYTHON_INSTALL_DIR "${CMAKE_INSTALL_PREFIX}/${FASTJET_PYTHON_INSTALL_DIR}")
   endif()
   message(STATUS "Python interface enabled and will be installed to ${FASTJET_PYTHON_INSTALL_DIR}")
 endif()
-add_subdirectory(example)  # for examples (gets built after all the rest)
+if (NOT FASTJET_PYINTERFACE_ONLY)
+  add_subdirectory(example)  # for examples (gets built after all the rest)
+endif()
 
 
 # generate the configure file
@@ -376,49 +387,52 @@ configure_file(fastjet-config.in fastjet-config)
 install(PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/fastjet-config DESTINATION ${CMAKE_INSTALL_BINDIR})
 
 
-#--------------------------------------------------------------------------
-# enable testing functionality
-enable_testing()
-add_test(
-  NAME test_compare
-  COMMAND ${PROJECT_SOURCE_DIR}/test-compare.sh
-)
-
-#--------------------------------------------------------------------------
-# for accessing fastjet from other CMake projects
-include(CMakePackageConfigHelpers)
+if(NOT FASTJET_PYINTERFACE_ONLY)
+  #--------------------------------------------------------------------------
+  # enable testing functionality
+  enable_testing()
+  if (UNIX OR (WIN32 AND MINGW))
+    add_test(
+      NAME test_compare
+      COMMAND ${PROJECT_SOURCE_DIR}/test-compare.sh
+    )
+  endif()
 
-# allow Fastjet to work with find_package
-export(EXPORT FastjetTargets
-  NAMESPACE fastjet::
-  FILE "${CMAKE_CURRENT_BINARY_DIR}/fastjetTargets.cmake"
-)
+  #--------------------------------------------------------------------------
+  # for accessing fastjet from other CMake projects
+  include(CMakePackageConfigHelpers)
 
-install(EXPORT FastjetTargets
-  NAMESPACE fastjet::
-  FILE fastjetTargets.cmake
-  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
-)
+  # allow Fastjet to work with find_package
+  export(EXPORT FastjetTargets
+    NAMESPACE fastjet::
+    FILE "${CMAKE_CURRENT_BINARY_DIR}/fastjetTargets.cmake"
+  )
 
-configure_package_config_file(${CMAKE_CURRENT_SOURCE_DIR}/Config.cmake.in
-  "${CMAKE_CURRENT_BINARY_DIR}/fastjetConfig.cmake"
-  INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
-  NO_SET_AND_CHECK_MACRO
-  NO_CHECK_REQUIRED_COMPONENTS_MACRO
+  install(EXPORT FastjetTargets
+    NAMESPACE fastjet::
+    FILE fastjetTargets.cmake
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
   )
 
-write_basic_package_version_file(
-  "${CMAKE_CURRENT_BINARY_DIR}/fastjetConfigVersion.cmake"
-  VERSION "${PROJECT_VERSION}"
-  COMPATIBILITY AnyNewerVersion
+  configure_package_config_file(${CMAKE_CURRENT_SOURCE_DIR}/Config.cmake.in
+    "${CMAKE_CURRENT_BINARY_DIR}/fastjetConfig.cmake"
+    INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
+    NO_SET_AND_CHECK_MACRO
+    NO_CHECK_REQUIRED_COMPONENTS_MACRO
   )
 
-install(FILES
-  ${CMAKE_CURRENT_BINARY_DIR}/fastjetConfig.cmake
-  ${CMAKE_CURRENT_BINARY_DIR}/fastjetConfigVersion.cmake
-  DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
+  write_basic_package_version_file(
+    "${CMAKE_CURRENT_BINARY_DIR}/fastjetConfigVersion.cmake"
+    VERSION "${PROJECT_VERSION}"
+    COMPATIBILITY AnyNewerVersion
   )
 
+  install(FILES
+    ${CMAKE_CURRENT_BINARY_DIR}/fastjetConfig.cmake
+    ${CMAKE_CURRENT_BINARY_DIR}/fastjetConfigVersion.cmake
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/fastjet
+  )
+endif()
 
 message(STATUS "${CONFIG_SUMMARY}")
 
diff --git a/pyinterface/CMakeLists.txt b/pyinterface/CMakeLists.txt
index 7016f179..4f46602d 100644
--- a/pyinterface/CMakeLists.txt
+++ b/pyinterface/CMakeLists.txt
@@ -9,7 +9,10 @@ else()
   set(DEV_MODULE Development.Module)
 endif()
 
-find_package(Python 3.8 COMPONENTS Interpreter ${DEV_MODULE} REQUIRED)
+set(CFRPMP_TEMP "${CMAKE_FIND_ROOT_PATH_MODE_PROGRAM}")
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM ONLY)
+find_package(Python 3.8 COMPONENTS ${DEV_MODULE} REQUIRED)
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM "${CFRPMP_TEMP}")
 
 set_property(SOURCE fastjet.i PROPERTY CPLUSPLUS ON)
 set_property(SOURCE fastjet.i PROPERTY SWIG_MODULE_NAME __init__)
@@ -29,8 +32,8 @@ endif()
 # Unix only, Windows is different
 set_property(TARGET fastjet_swig PROPERTY SUFFIX ".${Python_SOABI}${CMAKE_SHARED_MODULE_SUFFIX}")
 
-target_link_libraries(fastjet_swig PUBLIC fastjet)
-target_link_libraries(fastjet_swig PUBLIC fastjettools)
+target_link_libraries(fastjet_swig PUBLIC fastjet::fastjet)
+target_link_libraries(fastjet_swig PUBLIC fastjet::fastjettools)
 target_link_libraries(fastjet_swig PUBLIC Python::Module)
 
 #set_target_properties(fastjet_swig PROPERTIES INSTALL_RPATH "$ORIGIN/../../../../lib")
-- 
2.49.0

